File: VB/DS12188/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraScheduler
Imports System.Data.OleDb

Namespace DS12188
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			' TODO: This line of code loads data into the 'carsDBLabelDataSet11.Labels' table. You can move, or remove it, as needed.
			Me.labelsTableAdapter1.Fill(Me.carsDBLabelDataSet11.Labels)
			' TODO: This line of code loads data into the 'carsDBLabelDataSet.Cars' table. You can move, or remove it, as needed.
			Me.carsTableAdapter.Fill(Me.carsDBLabelDataSet.Cars)
			' TODO: This line of code loads data into the 'carsDBLabelDataSet.CarScheduling' table. You can move, or remove it, as needed.
			Me.carSchedulingTableAdapter.Fill(Me.carsDBLabelDataSet.CarScheduling)

			AddHandler carSchedulingTableAdapter.Adapter.RowUpdated, AddressOf carSchedulingTableAdapter_RowUpdated

			InitializeLabels()
		End Sub

		Private Sub InitializeLabels()
			Me.labelsTableAdapter1.Fill(Me.carsDBLabelDataSet11.Labels)

			Dim labels As DataTable = Me.carsDBLabelDataSet11.Labels

			If labels.Rows.Count = 0 Then
				Return
			End If
			schedulerControl1.Storage.Appointments.Labels.Clear()

			schedulerControl1.Storage.Appointments.Labels.BeginUpdate()
			For i As Integer = 0 To labels.Rows.Count - 1
				Dim color As Color = Color.FromArgb(Int32.Parse(labels.Rows(i).ItemArray(1).ToString()))
				Dim dislayName As String = labels.Rows(i).ItemArray(2).ToString()
				Dim menuCaption As String = labels.Rows(i).ItemArray(3).ToString()
				Dim aptLabel As New AppointmentLabel(color, dislayName, menuCaption)
				schedulerControl1.Storage.Appointments.Labels.Add(aptLabel)
			Next i
			schedulerControl1.Storage.Appointments.Labels.EndUpdate()
		End Sub

		Private Sub gridView1_RowUpdated(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.RowObjectEventArgs) Handles gridView1.RowUpdated
			labelsTableAdapter1.Update(carsDBLabelDataSet11)
			carsDBLabelDataSet11.AcceptChanges()
			InitializeLabels()
			schedulerControl1.Refresh()
		End Sub

		Private Sub OnApptChangedInsertedDeleted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs) Handles schedulerStorage1.AppointmentsChanged, schedulerStorage1.AppointmentsInserted, schedulerStorage1.AppointmentsDeleted
			carSchedulingTableAdapter.Update(carsDBLabelDataSet)
			carsDBLabelDataSet.AcceptChanges()
		End Sub

		Private Sub carSchedulingTableAdapter_RowUpdated(ByVal sender As Object, ByVal e As OleDbRowUpdatedEventArgs)
			If e.Status = UpdateStatus.Continue AndAlso e.StatementType = StatementType.Insert Then
				Dim id As Integer = 0
				Using cmd As New OleDbCommand("SELECT @@IDENTITY", carSchedulingTableAdapter.Connection)
					id = CInt(Fix(cmd.ExecuteScalar()))
				End Using
				e.Row("ID") = id
			End If
		End Sub

	End Class
End Namespace

File: CS/DS12188/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraScheduler;
using System.Data.OleDb;

namespace DS12188
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            // TODO: This line of code loads data into the 'carsDBLabelDataSet11.Labels' table. You can move, or remove it, as needed.
            this.labelsTableAdapter1.Fill(this.carsDBLabelDataSet11.Labels);
            // TODO: This line of code loads data into the 'carsDBLabelDataSet.Cars' table. You can move, or remove it, as needed.
            this.carsTableAdapter.Fill(this.carsDBLabelDataSet.Cars);
            // TODO: This line of code loads data into the 'carsDBLabelDataSet.CarScheduling' table. You can move, or remove it, as needed.
            this.carSchedulingTableAdapter.Fill(this.carsDBLabelDataSet.CarScheduling);

            carSchedulingTableAdapter.Adapter.RowUpdated += new OleDbRowUpdatedEventHandler (carSchedulingTableAdapter_RowUpdated);

            InitializeLabels();
        }

        private void InitializeLabels()
        {
            this.labelsTableAdapter1.Fill(this.carsDBLabelDataSet11.Labels);

            DataTable labels = this.carsDBLabelDataSet11.Labels;

            if (labels.Rows.Count == 0)
                return;
            schedulerControl1.Storage.Appointments.Labels.Clear();

            schedulerControl1.Storage.Appointments.Labels.BeginUpdate();
            for (int i = 0; i < labels.Rows.Count; i++)
            {
                Color color = Color.FromArgb(Int32.Parse(labels.Rows[i].ItemArray[1].ToString()));
                string dislayName = labels.Rows[i].ItemArray[2].ToString();
                string menuCaption = labels.Rows[i].ItemArray[3].ToString();
                AppointmentLabel aptLabel = new AppointmentLabel(color, dislayName, menuCaption);
                schedulerControl1.Storage.Appointments.Labels.Add(aptLabel);
            }
            schedulerControl1.Storage.Appointments.Labels.EndUpdate();
        }

        private void gridView1_RowUpdated(object sender, DevExpress.XtraGrid.Views.Base.RowObjectEventArgs e)
        {
            labelsTableAdapter1.Update(carsDBLabelDataSet11);
            carsDBLabelDataSet11.AcceptChanges();
            InitializeLabels();
            schedulerControl1.Refresh();
        }

        private void OnApptChangedInsertedDeleted(object sender, PersistentObjectsEventArgs e)
        {
            carSchedulingTableAdapter.Update(carsDBLabelDataSet);
            carsDBLabelDataSet.AcceptChanges();
        }

        private void carSchedulingTableAdapter_RowUpdated(object sender, OleDbRowUpdatedEventArgs e)
        {
            if (e.Status == UpdateStatus.Continue && e.StatementType == StatementType.Insert)
            {
                int id = 0;
                using (OleDbCommand cmd = new OleDbCommand("SELECT @@IDENTITY", carSchedulingTableAdapter.Connection))
                {
                    id = (int)cmd.ExecuteScalar();
                }
                e.Row["ID"] = id;
            }
        }

    }
}